datanommer.models/datanommer/models/__init__.py

```diff
@@ -282,15 +281,16 @@ def get_or_create(cls, name):
         Return the instance of the class with the specified name. If it doesn't
         already exist, create it.
         """
-        # Use an INSERT ... SELECT to guarantee we don't get unique constraint
-        # violations if multiple instances of datanommer are trying to insert the same
-        # value at the same time.
-        not_exists = ~exists(select([cls.__table__.c.name]).where(cls.name == name))
-        insert = cls.__table__.insert().\
-                 from_select([cls.__table__.c.name],
-                             select([literal(name)]).where(not_exists))
-        session.execute(insert)
-        return cls.query.filter_by(name=name).one()
+        try:
+            return cls.query.filter_by(name=name).one()
+        except NoResultFound:
+            try:
+                obj = cls(name=name)
+                session.add(obj)
+                return obj
+            except IntegrityError:
+                session.rollback()
```

Comment:
This will throw away the insert of the message that we did earlier in the transaction. For this to work, you'd need to move the conditional creation of the Users and Packages to before the message is added to the session. You would also need to re-add any Users and/or Packages that you added before the User or Pakcage that raised the IntegrityError. I don't think this is a workable solution in this case.